r/src/matrix_io.cpp

```diff
@@ -70,6 +70,11 @@ template <class T> List dims_matrix(StoredMatrix<T> &&mat, bool transpose) {
     }
 }
 
+template <class T> List dims_matrix(std::unique_ptr<MatrixLoader<T>> &&mat, bool transpose) {
+    // This is only safe because we know the main `dims_matrix` function doesn't store any references to the object
+    return dims_matrix(std::move(*mat), transpose);
+}
```

Comment:
I need to do a bit more research on the safety of this before merging. I was just running some small C++ tests and getting surprised by when destructors got called, so my mental model is not quite right

Reply:
I think I have figured this out now. The solution I used was:
- Make the base `dims_matrix` take an l-value reference, since that provides more ownership guarantees to the caller (i.e. the object will not get consumed by the function call)
- Make a helper overload that takes an r-value reference. This provides fewer guarantees to the caller but is otherwise safe
- Dereferencing a std::unique_ptr with `*` produces an l-value, which now can be passed directly to the l-value reference overload of `dims_matrix`